File: src/main/java/dev/strawn/frugal/FrugalParserDefinition.java
package dev.strawn.frugal;

import com.intellij.lang.ASTNode;
import com.intellij.lang.ParserDefinition;
import com.intellij.lang.PsiParser;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.project.Project;
import com.intellij.psi.FileViewProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IFileElementType;
import com.intellij.psi.tree.TokenSet;
import org.jetbrains.annotations.NotNull;

public class FrugalParserDefinition implements ParserDefinition {
    
    public static final IFileElementType FILE = new IFileElementType(FrugalLanguage.INSTANCE);
    public static final TokenSet WHITE_SPACES = TokenSet.create(TokenType.WHITE_SPACE);
    public static final TokenSet COMMENTS = TokenSet.EMPTY;
    public static final TokenSet STRINGS = TokenSet.EMPTY;
    
    @Override
    public @NotNull Lexer createLexer(Project project) {
        return new FrugalLexer();
    }
    
    @Override
    public @NotNull PsiParser createParser(Project project) {
        return new FrugalParser();
    }
    
    @Override
    public @NotNull IFileElementType getFileNodeType() {
        return FILE;
    }
    
    @Override
    public @NotNull TokenSet getCommentTokens() {
        return COMMENTS;
    }
    
    @Override
    public @NotNull TokenSet getStringLiteralElements() {
        return STRINGS;
    }
    
    @Override
    public @NotNull TokenSet getWhitespaceTokens() {
        return WHITE_SPACES;
    }
    
    @Override
    public @NotNull PsiElement createElement(ASTNode node) {
        return new FrugalPsiElement(node);
    }
    
    @Override
    public @NotNull PsiFile createFile(@NotNull FileViewProvider viewProvider) {
        return new FrugalFile(viewProvider);
    }
}

File: src/main/kotlin/dev/strawn/frugal/FrugalParserDefinition.kt
package dev.strawn.frugal

import com.intellij.lang.ASTNode
import com.intellij.lang.ParserDefinition
import com.intellij.lang.PsiParser
import com.intellij.lexer.Lexer
import com.intellij.openapi.project.Project
import com.intellij.psi.FileViewProvider
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.TokenType
import com.intellij.psi.tree.IFileElementType
import com.intellij.psi.tree.TokenSet

class FrugalParserDefinition : ParserDefinition {
    
    companion object {
        val FILE = IFileElementType(FrugalLanguage)
        val WHITE_SPACES = TokenSet.create(TokenType.WHITE_SPACE)
        val COMMENTS = TokenSet.EMPTY
        val STRINGS = TokenSet.EMPTY
    }
    
    override fun createLexer(project: Project?): Lexer {
        return FrugalLexer()
    }
    
    override fun createParser(project: Project?): PsiParser {
        return FrugalParser()
    }
    
    override fun getFileNodeType(): IFileElementType {
        return FILE
    }
    
    override fun getCommentTokens(): TokenSet {
        return COMMENTS
    }
    
    override fun getStringLiteralElements(): TokenSet {
        return STRINGS
    }
    
    override fun getWhitespaceTokens(): TokenSet {
        return WHITE_SPACES
    }
    
    override fun createElement(node: ASTNode?): PsiElement {
        return FrugalPsiElement(node)
    }
    
    override fun createFile(viewProvider: FileViewProvider): PsiFile {
        return FrugalFile(viewProvider)
    }
}